Fix payment method section missing for Affirm and Afterpay on transaction details page

mgascam opened the pull request:
Fixes #7343 

#### Changes proposed in this Pull Request

<!--
Title: A descriptive, yet concise, title.
-->
This pull request adds the payment method section to the details page of a transaction paid with Affirm or Afterpay.
<!--
Description: Write a brief summary about this PR. As you compose your summary, consider each of these questions and address them if appropriate. Why is this change needed? What does this change do? Were there other solutions you considered? Why did you choose to pursue this solution? Describe any trade-offs you might have had to make.
-->

<!--
Questions for PR author:
- How can this code break?
- What are we doing to make sure this code doesn't break?
-->

<!--
Images or gifs: Include before and after screenshots or gifs/videos when it makes sense.
-->

### Affirm

**Before**

<img width="1154" alt="Screenshot 2023-10-04 at 16 25 24" src="https://github.com/Automattic/woocommerce-payments/assets/1553182/b10ea3c2-b568-4adc-8756-713ca9fb0630">

**After**

<img width="1154" alt="Screenshot 2023-10-04 at 16 22 46" src="https://github.com/Automattic/woocommerce-payments/assets/1553182/eb0c0753-2d59-4d2d-aaf5-ebdb9bea7344">


### Afterpay

**Before**

<img width="1154" alt="Screenshot 2023-10-04 at 16 26 33" src="https://github.com/Automattic/woocommerce-payments/assets/1553182/6d6b9a0c-cec9-4a4f-bfdc-94b2483014b3">

**After**

<img width="1154" alt="Screenshot 2023-10-04 at 16 25 51" src="https://github.com/Automattic/woocommerce-payments/assets/1553182/e554f8d4-510a-4b20-b253-aff06ea06d6f">

#### Testing instructions

<!--
Testing instructions: How should this be tested and how can a reviewer test the end-user functionality? Are there known issues that you plan to address in a future PR? Are there any side effects that readers should be aware of?
-->

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

* You'll need at least one Affirm and Afterpay transactio.
* Go to WP Admin > Payments > Transactions
* Locate one Affirm transaction and click it to go the details page
* Scroll to bottom and check that the payment method section appears.
* Repeat the process with one Afterpay transaction
-------------------

- [ ] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [ ] Covered with tests (or have a good reason not to test in description ☝️)
- [ ] Tested on mobile (or does not apply)

**Post merge**

<!--
Make sure you edit the page for the current release when adding testing instructions.
We often create a blank page ahead of time for the next release.
If this PR need not be QA tested, edit to 'QA Testing Not Applicable'
-->

- [x] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) following [these instructions](https://github.com/Automattic/woocommerce-payments/wiki/How-to-write-good-manual-testing-scenarios) :  'QA Testing Not Applicable'
- [ ] Add or update [critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Critical-flows) and [testing instructions for critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Testing-instructions-for-critical-flows), if applicable.
- [ ] Add what's changed (description, screenshot, demo videos etc.) to the release announcement post, if applicable.

mgascam committed:
Add  payment method section for Affirm

mgascam committed:
Add payment method section for Afterpay

mgascam committed:
Merge branch 'develop' into fix/7343-missing-payment-details-section-in-payment-details-page-for-affirm-and-afterpay-payment-methods

botwoo commented:
### Test the build 

#### Option 1. Jetpack Beta

- Install and activate [Jetpack Beta](https://jetpack.com/download-jetpack-beta/).
- Use this build by searching for PR number `7345` or branch name `fix/7343-missing-payment-details-section-in-payment-details-page-for-affirm-and-afterpay-payment-methods` in your-test.site/wp-admin/admin.php?page=jetpack-beta&plugin=woocommerce-payments          

#### Option 2. Jurassic Ninja - available for logged-in A12s

:rocket: [Launch a JN site with this branch](https://jurassic.ninja/create/?jetpack-beta&shortlived&nojetpack&woocommerce&woocommerce-payments-dev-tools&branches.woocommerce-payments=fix/7343-missing-payment-details-section-in-payment-details-page-for-affirm-and-afterpay-payment-methods) :rocket:

:information_source: Install this [Tampermonkey script](https://github.com/Automattic/woocommerce-payments/tree/develop/bin/wcpay-live-branches) to get more options.

---

Build info: 

- Latest commit: bb32d39d4bdd670de490df590bc71e3e091954a2
- Build time: 2023-10-05 11:56:52 UTC

_Note: the build is updated when a new commit is pushed to this PR._

mgascam committed:
Merge branch 'develop' into fix/7343-missing-payment-details-section-in-payment-details-page-for-affirm-and-afterpay-payment-methods

mgascam committed:
Add changelog.

dmvrtx commented on client/payment-details/payment-method/index.js:
```diff
@@ -20,6 +20,8 @@ import IdealDetails from './ideal';
 import P24Details from './p24';
 import SepaDetails from './sepa';
 import SofortDetails from './sofort';
+import AffirmDetails from './affirm';
+import AfterpayClearpayDetails from './afterpay-clearpay';
 
```
If we sort imports here we will avoid conflicts with #7282

mgascam replied:
Done here https://github.com/Automattic/woocommerce-payments/pull/7345/commits/07c70fac9f49ea458d328ddb925816251f2f49a9 thanks!

dmvrtx reviewed (approved):
Thank you for addressing this @mgascam. Provisionally approved with a minor comment. Tested on my local environment, looks good.

mgascam committed:
PR feedback: sort imports

kalessil reviewed (commented):
Thanks for adding screenshots @mgascam !

Some remarks:
- I'd expect the new components to be in TS (this can be done in a separate PR if this one is aiming before code freeze)
- It'd be better to have the same set of fields (in Affirm case it's less, so we could follow the same approach)
- I'm a bit confused as don't understand where do we invoke the new components

kalessil commented:
My comments are not show-stoppers but subjects to consider for us and learn something new for me.

mgascam committed:
PR feedback: remove fields for Afterpay

mgascam commented:
Thanks @kalessil , answering your questions below:

> * I'd expect the new components to be in TS (this can be done in a separate PR if this one is aiming before code freeze)

Yes,  it would be great to include this change in next release so I'll give it a try (timeboxed) to perform the migration in this PR. If I don't progress quick enough, I'll tackle it in different PR.

> * It'd be better to have the same set of fields (in Affirm case it's less, so we could follow the same approach)

Indeed, I added the additional fields for Afterpay because Stripe docs [list](https://stripe.com/docs/api/charges/object#charge_object-payment_method_details-afterpay_clearpay) those as possible fields. But truth is those fields don't add much information (we already have the order in this view) so I removed them in this commit https://github.com/Automattic/woocommerce-payments/pull/7345/commits/8240ccae1bcda0a713ee76d5b56995a4d57ef1f3

> * I'm a bit confused as don't understand where do we invoke the new components

The `PaymentDetailsPaymentMethod` is the one that renders the components. It holds a dictionary of all available `PaymentMethodDetails`. It looks up the type informed in the `payment_method_details` of the `charge` in the dictionary and renders it or returns null. It happens here: https://github.com/Automattic/woocommerce-payments/blob/07c70fac9f49ea458d328ddb925816251f2f49a9/client/payment-details/payment-method/index.js#L50

mgascam committed:
Merge branch 'develop' into fix/7343-missing-payment-details-section-in-payment-details-page-for-affirm-and-afterpay-payment-methods

mgascam committed:
Merge branch 'develop' into fix/7343-missing-payment-details-section-in-payment-details-page-for-affirm-and-afterpay-payment-methods

mgascam committed:
Add dash fallback for undefined details

mgascam committed:
Merge branch 'develop' into fix/7343-missing-payment-details-section-in-payment-details-page-for-affirm-and-afterpay-payment-methods

mgascam commented:
Sent to the merge queue. 

I created this https://github.com/Automattic/woocommerce-payments/issues/7388 to address the https://github.com/Automattic/woocommerce-payments/pull/7345#issuecomment-1747375749 to TSX in a different PR.
 
FYI: @kalessil
